mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java

```diff
@@ -688,10 +697,23 @@ public MessageResultIterator getMessages(MessageRange set, FetchGroup fetchGroup
 
     }
 
-    protected Map<MessageUid, MessageMetaData> deleteMarkedInMailbox(final MessageRange range, MailboxSession session) throws MailboxException {
-
+    protected Map<MessageUid, MessageMetaData> deleteWithPreHooks(final MessageRange range, MailboxSession session) throws MailboxException {
         final MessageMapper messageMapper = mapperFactory.getMessageMapper(session);
+        List<MailboxMessage> messagesWillBeDeleted = ImmutableList.copyOf(
+            messageMapper.findInMailbox(mailbox, range, FetchType.Metadata, -1));
+
+        DeleteOperation deleteOperation = DeleteOperation.from(
+            messagesWillBeDeleted.stream()
+                .map(mailboxMessage -> MetadataWithMailboxId.from(mailboxMessage.metaData(), mailboxMessage.getMailboxId()))
+                .collect(Guavate.toImmutableList()));
+
+        return Flux.fromIterable(preDeletionHooks)
+            .flatMap(hook -> hook.notifyDelete(deleteOperation))
```

Comment:
Deleting a huge range of mails on a configuration with several hooks will have an impact on deletion.

Reply:
what do you mean `impact`? are you mentioning about performance issue? I could place `.parallel()` in flux pipeline to optimize hooks execution if you agree?

Reply:
It's just a reminder, I don't have a solution; and we don't have a problem yet ;)

There are multiple solution for such problems, wait and see.

Reply:
> Deleting a huge range of mails on a configuration with several hooks will have an impact on deletion.

This is necessary: we need hook execution to be synchronous. That's why it's hooks and not listeners.